test_conformance/subgroups/workgroup_common_templates.h

```diff
@@ -0,0 +1,585 @@
+                            break;
+                        }
+                    }
+                    if (lowest_active_id == -1)
+                    {
+                        log_error(
+                            "ERROR: sub_group_broadcast_first(%s) do not found "
+                            "any matching values in sub group %d in group %d\n",
+                            TypeManager<Ty>::name(), j, k);
+                        return -1;
+                    }
+                    for (i = 0; i < n; ++i)
+                    {
+                        tr = mx[ii
+                                + lowest_active_id]; //  findout if broadcasted
+                                                     //  value is the same
+                        rr = my[ii + i]; //  findout if broadcasted to all
+                        if (!compare(rr, tr))
+                        {
+                            log_error(
+                                "ERROR: sub_group_broadcast_first(%s) mismatch "
+                                "for local id %d in sub group %d in group %d\n",
+                                TypeManager<Ty>::name(), i, j, k);
+                        }
+                    }
+                }
+                else
+                {
+                    for (i = 0; i < n; ++i)
+                    {
+                        if (Which == 2 && i >= NON_UNIFORM)
+                        {
+                            break; // non uniform case - only first 4 workitems
+                                   // should broadcast. Others are undefined.
+                        }
+                        rr = my[ii + i]; // read device outputs for work_item in
+                                         // the subgroup
+                        if (!compare(rr, tr))
+                        {
+                            if (Which == 0)
+                            {
+                                log_error("ERROR: sub_group_broadcast(%s) "
+                                          "mismatch for local id %d in sub "
+                                          "group %d in group %d\n",
+                                          TypeManager<Ty>::name(), i, j, k);
+                            }
+                            if (Which == 2)
+                            {
+                                log_error("ERROR: "
+                                          "sub_group_non_uniform_broadcast(%s) "
+                                          "mismatch for local id %d in sub "
+                                          "group %d in group %d\n",
+                                          TypeManager<Ty>::name(), i, j, k);
+                            }
+                            return -1;
+                        }
+                    }
+                }
+            }
+            x += nw;
+            y += nw;
+            m += 4 * nw;
+        }
+        return 0;
+    }
+};
+
+template <typename Ty, int Which> struct OPERATION;
```

Comment:
All these templates should be condensed into a switch on "Which". Let the compiler constant-propagate and remove the unused cases.
Would make this far simpler to understand and shorter.